src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseHybridFVM.cpp

```diff
@@ -398,109 +398,124 @@ void SinglePhaseHybridFVM::saveAquiferConvergedState( real64 const & time,
-                                                            faceDofNumber.toNestedViewConst(),
-                                                            faceGhostRank.toNestedViewConst(),
-                                                            elemRegionList.toNestedViewConst(),
-                                                            elemSubRegionList.toNestedViewConst(),
-                                                            elemList.toNestedViewConst(),
-                                                            flowAccessors.get( extrinsicMeshData::elementVolume{} ),
-                                                            defaultViscosity,
-                                                            &localResidualNorm[3] );
+      ResidualNormKernelFactory::
+      createAndLaunch< parallelDevicePolicy<> >( m_normType,
+                                                 rankOffset,
+                                                 faceDofKey,
+                                                 localRhs,
+                                                 m_regionFilter.toViewConst(),
+                                                 getName(),
+                                                 elemManager,
+                                                 faceManager,
+                                                 defaultViscosity,
+                                                 dt,
+                                                 faceResidualNorm,
+                                                 faceResidualNormalizer );
+
+    // step 2.2: reduction across meshBodies/regions/subRegions
+
+    if( m_normType == solverBaseKernels::NormType::Linf )
+    {
+      if( faceResidualNorm[0] > localResidualNorm )
+      {
+        localResidualNorm = faceResidualNorm[0];
+      }
+    }
+    else
+    {
+      localResidualNorm += faceResidualNorm[0];
+      localResidualNormalizer += faceResidualNormalizer[0];
+    }
+  } );
 
+  // step 3: second reduction across MPI ranks
 
-  } );
-  // 3. Combine the two norms
-
-  // compute global residual norm
-  MpiWrapper::allReduce( localResidualNorm,
-                         globalResidualNorm,
-                         4,
-                         MPI_SUM,
-                         MPI_COMM_GEOSX );
-
-  real64 const elemResidualNorm = sqrt( globalResidualNorm[0] )
-                                  / ( ( globalResidualNorm[1] + m_fluxEstimate ) / (globalResidualNorm[2]+1) );
-  real64 const faceResidualNorm = sqrt( globalResidualNorm[3] );
-
-  real64 const residualNorm = ( elemResidualNorm > faceResidualNorm )
-                            ? elemResidualNorm
-                            : faceResidualNorm;
+  real64 const residualNorm = ( m_normType == solverBaseKernels::NormType::Linf )
+    ? MpiWrapper::max( localResidualNorm )
+    : sqrt( MpiWrapper::sum( localResidualNorm ) ) / MpiWrapper::sum( localResidualNormalizer );
```

Comment:
Same here about OOP.

Reply:
Here as well, I put this code in a helper function.